[OpenXR] Fallback to use GBM when EGL extension MESA_image_dma_buf_export is not available

carlosgcampos opened the pull request:
#### 8829cb72321e001331b5a5bee6c48e195fd4b353
<pre>
[OpenXR] Fallback to use GBM when EGL extension MESA_image_dma_buf_export is not available
<a href="https://bugs.webkit.org/show_bug.cgi?id=297717">https://bugs.webkit.org/show_bug.cgi?id=297717</a>

Reviewed by Adrian Perez de Castro.

We are currently using MESA_image_dma_buf_export unconditionally. When
not available we can use GBM to allocate our own buffer for the web
process to render into and then we copy it to the openxr texture on end
frame.

Canonical link: <a href="https://commits.webkit.org/299044@main">https://commits.webkit.org/299044@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af89d0c38df45c5b3ee3a4f157fef2171b34e792

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117456 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37126 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27746 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123556 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69447 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37822 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45713 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89136 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/43793 "An unexpected error occured. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120408 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30109 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105318 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69645 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29165 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67230 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99514 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23614 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126676 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44353 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33344 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97802 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44711 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101552 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97596 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42964 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20905 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/40704 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18768 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44226 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49885 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43683 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47027 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45378 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

spenap commented on Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp:
```diff
@@ -54,17 +71,11 @@ std::optional<PlatformXR::FrameData::ExternalTexture> OpenXRLayer::exportOpenXRT
 
-    auto* glContext = WebCore::GLContext::current();
-    ASSERT(glContext);
+    auto image = display.createImage(context.platformContext(), EGL_GL_TEXTURE_2D, (EGLClientBuffer)(uint64_t)openxrTexture, { });
```
I see you're moving existing code, but we might consider expressing this as

```suggestion
    auto image = display.createImage(context.platformContext(), EGL_GL_TEXTURE_2D, reinterpret_cast<EGLClientBuffer>(static_cast<uintptr_t>(openxrTexture)), { });
```

I've grep'd the code and this approach is used in ANGLE, bmalloc or JavaScriptCore (where they use `void*` instead of `EGLClientBuffer`, but it'd be semantically the same)

carlosgcampos replied:
I don't know, but I prefer to change this in a separate patch and do the same in AcceleratedSurface.cpp

aperezdc replied:
We have also had instances in the past where, no matter what we did with C++ casts, there would be always some odd platform where the code would not compile, and ended up using a plain C cast (for examples, see [263299@main](https://commits.webkit.org/263299@main), or [194561@main](https://commits.webkit.org/194561@main))&mdash;I don't know if that would be the case here, only wanting to point out there are precedents where the most reasonable thing to do was the good ol' C cast.

spenap commented on Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp:
```diff
@@ -74,22 +85,22 @@ std::optional<PlatformXR::FrameData::ExternalTexture> OpenXRLayer::exportOpenXRT
 
-    display->destroyImage(image);
+    display.destroyImage(image);
 
-    releaseTextureOnError.release();
+    releaseImageOnError.release();
```
If I understand this correctly, rather than allowing the `ScopeExit` object to run automatically when leaving the method, we're manually invoking its _exitFunction_ and then releasing, so it doesn't do it again.

Isn't that equivalent to letting the `ScopeExit` object get destroyed at the end of the function? We could rename it so rather than `releaseImageOnError` it becomes `displayImageCleanup`

carlosgcampos replied:
Yes, we could do that if we set the image to nullptr after explicitly destroying it. Again, out of scope of this patch.

spenap replied:
(not disputing it can be left out for later, but) my point is that unless we want to destroy the image exactly at this point because of a side effect, it is perfectly fine not to destroy it explicitly, and allow the scope object to do it as we leave the method

carlosgcampos replied:
Ah you mean not explicitly release the image and always use the scope exit. Yes, that should work too.

spenap commented on Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp:
```diff
@@ -110,6 +121,155 @@ std::optional<PlatformXR::FrameData::ExternalTexture> OpenXRLayer::exportOpenXRT
+        EGL_LINUX_DRM_FOURCC_EXT, static_cast<EGLAttrib>(fourcc),
+    };
+
+#define ADD_PLANE_ATTRIBUTES(planeIndex) { \
```
This comment is out of scope for this PR and probably for a while, but I notice we have 7/8 definitions of `ADD_PLANE_ATTRIBUTES` that do a very similar thing but have some minor differences.

I wonder if there is a way to consolidate it elsewhere.

aperezdc reviewed (approved):
LGTM.

carlosgcampos committed:
[OpenXR] Fallback to use GBM when EGL extension MESA_image_dma_buf_export is not available
https://bugs.webkit.org/show_bug.cgi?id=297717

Reviewed by Adrian Perez de Castro.

We are currently using MESA_image_dma_buf_export unconditionally. When
not available we can use GBM to allocate our own buffer for the web
process to render into and then we copy it to the openxr texture on end
frame.

Canonical link: https://commits.webkit.org/299044@main

svillar commented on Source/WebKit/UIProcess/XR/openxr/OpenXRLayer.cpp:
```diff
@@ -145,8 +305,10 @@ std::optional<PlatformXR::FrameData::LayerData> OpenXRLayerProjection::startFram
     auto externalTexture = exportOpenXRTexture(*texture);
-    if (!externalTexture)
+    if (!externalTexture) {
+        m_swapchain->releaseImage();
```
Arriving late, but this is not correct. The release should happen in endFrame. Even if startFrame fails at this point we have to call endFrame anyway, and there is where the release should happen.

carlosgcampos replied:
That release was there before, but done in the scope exit when we failed to create the EGLImage, so when nullopt was returned by exportOpenXRTexture. I moved it here to keep that behavior and release it also for the case of failing to export using GBM. I don't know if that was wrong or not, but I don't think this patch changed the behavior, unless I'm missing something...
